### ISSUE TITLE
export C API

### ISSUE THREAD

**Issue by lewisxy** [rclone/rclone#4891]
I would like to integrate rclone features into my own application. One way to do this is to build rclone as a static (`-buildmode=c-archive`) or dynamic (`-buildmode=c-shared`) library, and link against it. However, both of these approach require specifying the functions to export via `//export` comment (see [go documentation](https://golang.org/cmd/go/#hdr-Build_modes)). I am not sure whether there are plan to support this officially (it would be great if so). If not, could you provide some hints of how to implement this in term of the internal design of rclone? Thanks.

**Comment by ncw** [comment]
There are lots of interfaces internally in rclone.

Rclone was designed to be used from other programs either by executing it as a command line program, or by using the [remote control API](https://rclone.org/rc/). The remote control API is a stable interface and involves HTTP POST of JSON blobs.

So you could run rclone as a subprocess and do just that.

It would be nice if the API supported running over stdin/stdout like for example `rclone serve restic` does. This uses HTTP2 over stdin/stdout to multiplex calls.

If you wanted a slightly tighter intergration then the easiest thing to do would be to export the remote control interface. So instead of passing JSON blobs back and forth over the API just pass serialized JSON strings to an exported function and get serialized JSON strings in return.

Serializing things to JSON will save a world of pain mapping things like Go maps between APIs even if it isn't the most efficient way of doing things.

I plan to use this technique to join rclone compiled as wasm to the webgui frontend (JS) at some point so you can have a totally in-browser rclone.

What language do you want to integrate rclone into?

**Comment by lewisxy** [comment]
> What language do you want to integrate rclone into?

I would like to create an IOS application that use rclone feature, so probably swift. The tricky part is that IOS applications are not allowed to run multiple processes. So it's not possible (or very difficult) to start a server at background and then perform remote control call to it. That's why I am looking for building rclone as a library so that it can be run in a single process. There might be some other use cases that will benefit from such library.

> If you wanted a slightly tighter intergration then the easiest thing to do would be to export the remote control interface. So instead of passing JSON blobs back and forth over the API just pass serialized JSON strings to an exported function and get serialized JSON strings in return.

I would definitely give this a try. Thanks for the advice.

**Comment by ncw** [comment]
You'll probably need to write a small amount of Go code (or I can help you with that) which is something like

    func CRPC(in string) (out string, int status)

where `in` is a serialized json blob, `out` is the same and `status` is the return code of the call - 0 for OK, non zero for error. Or maybe it should be HTTP status codes.

That will let you do everything the remote control API does, with the exception of reading data from an object and writing data to an object. You can copy objects to and from local files but not into memory with this interface.

If that is something you want to do then that will need another interface - shouldn't be too hard though.

**Comment by lewisxy** [comment]
After spending some time on this, I managed to export an API that works, see [here](https://github.com/lewisxy/rclone-library/tree/master/exports). Use the Makefile to build the library and C program.

Things that I am not sure how to do yet.
1. When the C program executes, the go runtime does not automatically initialize itself until the first go function is called, and I am not sure when will the go runtime terminates. So I added 2 additional function, `Cinit` and `Cdestroy`, but I am not sure what to put in there (or if they are necessary at all).
2. So far, the code I used primarily come from [`handlePost` in `rcserver.go`](https://github.com/rclone/rclone/blob/master/fs/rc/rcserver/rcserver.go#L231). The commands that does not work are those require request/response (those with `NeedsRequest` or `NeedsResponse` in registry). Fortunately, most commands does not need these, so it's probably fine not to implement those. Adding support to those can be slightly challenging.
3. I am not sure how to load config file. I am aware there are commands that modify the config file, but I am not sure how should to specify which file to use as config (the current library will use the system default at `~/.config/rclone/rclone.conf`). It would be nice if config can be loaded from this API (either specify the filename or just pass as string).

**Comment by ncw** [comment]
> After spending some time on this, I managed to export an API that works, see [here](https://github.com/lewisxy/rclone-library/tree/master/exports). Use the Makefile to build the library and C program.

Well done - that looks good :-)

> Things that I am not sure how to do yet.
> 
> 1. When the C program executes, the go runtime does not automatically initialize itself until the first go function is called, and I am not sure when will the go runtime terminates. So I added 2 additional function, `Cinit` and `Cdestroy`, but I am not sure what to put in there (or if they are necessary at all).

I didn't know that about the go runtime not initializing until the first call. I would guess that it never really finalizes itself - it wasn't designed for finalization.

> 2. So far, the code I used primarily come from [`handlePost` in `rcserver.go`](https://github.com/rclone/rclone/blob/master/fs/rc/rcserver/rcserver.go#L231). The commands that does not work are those require request/response (those with `NeedsRequest` or `NeedsResponse` in registry). Fortunately, most commands does not need these, so it's probably fine not to implement those. Adding support to those can be slightly challenging.

I think you can safely ignore those.

> 3. I am not sure how to load config file. I am aware there are commands that modify the config file, but I am not sure how should to specify which file to use as config (the current library will use the system default at `~/.config/rclone/rclone.conf`). It would be nice if config can be loaded from this API (either specify the filename or just pass as string).

Good point... I think you probably want to call an rclone initialization function `config.LoadConfig(ctx)` if you want to use the standard rclone config file. You can set which config file to use by setting `config.ConfigPath` - do that before the `LoadConfig` call.

You can use your own config system if you want by overriding the function pointers fs.ConfigFileGet and fs.ConfigFileSet but it is probably easiest to use the rclone config file.

**Comment by ivandeex** [comment]
I can see that all questions are answered.
Is there anything we want to add in the rclone docs.
If not, I think we can close this one.

**Comment by lewisxy** [comment]
I was quite busy at the moment and don't have time or plan to work on this for the near future (probably for few months). So I guess I will close it. Thanks for all the replies. They were helpful. I will probably open new issues when I run into other problems in the future.

**Comment by ncw** [comment]
@lewisxy I'd quite like to make your code the official C-API for rclone.

Would you mind if I pulled your initial work into rclone and then worked on it some more?

What I'd like to do is cherry pick your commits from your tree, squash them and tidy them and that can be the initial version. I'll then add more work in further commits. This would then get you a mention in the AUTHORS file.

I'm going to re-open this as a place for this feature to go.

**Comment by lewisxy** [comment]
Please feel free to use the code I have. I am glad that there is plan for official C API support and happy to be part of it.

**Comment by ncw** [comment]
Thank you @lewisxy and thank you very much for the idea and the initial implementation.

I think this will be really popular.

I've taken your work and moved it about a bit so it is in the `librclone` directory with the C test program in `ctest` . `go build` tries to build the `.c` program and link it so the build of rclone fails. I've renamed the exports so they all start with `Rclone` to name space them a bit and also make them exported to Go so they appear in the Go docs.

I've pushed the combined work to the [librclone](https://github.com/rclone/rclone/tree/librclone) branch
